### PR TITLE
Rewrite boilerplate impls into macro

### PR DESCRIPTION
**Pull request by ImmemorConsultrixContrarie**
Two problems:
1) I accidentally rustfmt it;
2) Usize type now has TYPENAME "usize" instead of "u32" or "u64". It could get back by implementing this trait for usize as a standalone with some "cfg!()"s.

### CODE REVIEW

**Comment by ImmemorConsultrixContrarie** [comment]
Oops... Actually, I wanted to make a draft PR, as it's not really complete: some comments should be deleted and some things in this code look really strange to me. For example, why do you need all this "as_slice", "from_slice", if you can just write
```
let mut buf = ToType::some_default_zero();
let Vsize = mem::size_of::<ValueType>();
let Tsize = mem::size_of::<ToType>();
unsafe {
  if Tsize > Vsize {
    // Some endian check, dunno if it should be cfged or checked in runtime.
    if cfg!(target_endian = "little") {
      ptr::copy_nonoverlapping(&value as *const u8, &mut buf as *mut u8, Vsize);
    } else {
      ptr::copy_nonoverlapping((&value as *const u8).add(Tsize-Vsize), &mut buf  as *mut u8, Vsize);
    }
  } else {
    // Tsize <= Vsize
    if cfg!(target_endian = "little") {
      ptr::copy_nonoverlapping(&value as *const u8, &mut buf as *mut u8, Tsize);
    } else {
      ptr::copy_nonoverlapping((&value as *const u8).add(Vsize-Tsize), &mut buf  as *mut u8, Vsize);
    }
  }
}
return buf;
```

File: src/fields.rs [hunk 1]
```diff
@@ -899,180 +945,211 @@ The result of transforming `value as U`. Where `U` is wider than `T`, this
 zero-extends; where `U` is narrower, it truncates.
 **/
 fn resize<T, U>(value: T) -> U
```
Comment (ImmemorConsultrixContrarie) [inline src/fields.rs]: That's the only notable change actually. Big-endian should be tested, but I don't think there would be any problems.

File: src/fields.rs [hunk 2]
```diff
@@ -899,180 +945,211 @@ The result of transforming `value as U`. Where `U` is wider than `T`, this
...
+                    &value as *const T as *const u8,
+                    &mut buf as *mut U as *mut u8,
+                    size_of_from,
+                );
+            }
+        }
+    }
+
+    buf
 }
```
Comment (ImmemorConsultrixContrarie) [inline src/fields.rs]: That's the only notable change actually. Big-endian should be tested, but I don't think there would be any problems.

**Comment by ImmemorConsultrixContrarie** [comment]
Travis failed with "could not find `stringify` in `core`"... This macro is stable since 1.0.0, wtf?

File: src/fields.rs [hunk 3]
```diff
@@ -899,180 +945,211 @@ The result of transforming `value as U`. Where `U` is wider than `T`, this
...
+    let size_of_into = mem::size_of::<U>();
+    unsafe {
+        if cfg!(target_endian = "big") {
+            // Big endian fills from the end.
+            if size_of_from > size_of_into {
+                // T has more bytes, move its pointer to match the length of U.
+                copy_nonoverlapping(
+                    (&value as *const T as *const u8).add(size_of_from - size_of_into),
+                    &mut buf as *mut U as *mut u8,
+                    size_of_from,
```
Comment (myrrlyn) [inline src/fields.rs]: This should be `size_of_into`; otherwise, the copy overruns the `buf` value.

Comment (ImmemorConsultrixContrarie) [reply src/fields.rs]: Yeah, typed it wrong.

**Comment by ImmemorConsultrixContrarie** [comment]
Btw, why not simply make 3-word wide BitMut

```rust
pub struct BitMut<'a, O, T>
where
    O: BitOrder,
    T: 'a + BitStore,
{
    /// A reference to a single bit in memory.
    pub(super) slot: &'a mut BitSlice<O, T>,
    /// A local cache for `Deref` usage.
    pub(super) data: bool,
}
```

into something like this (two words wide):

```rust
pub struct BitMut<'a, O, T>
where
    O: BitOrder,
    T: 'a + BitStore,
{
    // Tell the compiler, this struct holds mutable access of the slice.
    _slice: core::marker::PhantomData<&'a mut BitSlice<O, T>>,
    // Writer.
    word: AtomicPtr<T>,
    // Derefed data, writes on drop as it does now.
    data: bool,
    // A mask-creator.
    place: u8, // or `u16`, but until `u256` would be a thing, u16 is overkill
}
```

Or make "split_mut" create a tuple of (&mut self, BitSliceWithOneClonedStorage), so this thing doesn't even need atomics, won't has atomics overhead, and could be Send at all times.

**Comment by myrrlyn** [comment]
Thank you for the PR. I appreciate your willingness to find improvements as well as defects. Other than the copy-size note, I have nothing to really put in the review workflow for the contents of the changes, but here's what I have for the PR as a whole:

## `src/fields.rs`

My `fields::resize` implementation was a second draft, and I stopped working on it as soon as I got to a useful point. The pointer-arithmetic version you wrote is cleaner, has no annoying edge cases in source code, and removes the detritus I had added to `BitStore` in order to support byte-based resizing. With the right copy length, it passes the test suite on my machine, and I'm happy to include it.

## `src/store.rs`

I also appreciate the macro collapse of the `store` trait implementations. I don't know why the path `core::stringify!` doesn't resolve, but as the macro is in the prelude of both `core` and `std`, it doesn't need the prefix in either `no_std` or `std` modes.

I've removed the `all_bits_zero` constructor function, as it is equivalent to the `FALSE` constant. `fields::resize` uses `U::FALSE` as the constructor instead of `U::all_bits_zero`, with identical behavior.

I am ambivalent on whether `usize` should rename to a fixed-size integer or not; the `TYPENAME` constant is now only used in formatting, the rename probably does not need to occur. I, personally, prefer fixed-size type names in the debug information, but I don't think it matters enough to warrant a change.

## Project Environment

I apologize for the *extremely* out-of-`rustfmt` style of the repository, and the lack of a `rustfmt.toml` to make `rustfmt` behave. This is the second contribution I've had, and the first from someone I didn't personally know and talk to out-of-band, so I just haven't set one up. I'll set up a `rustfmt.toml` and run `cargo fmt` on the repository in the near future, so that future contributors don't run into the hassle you did dealing with the formatting noise.

## Resolution

I've reshaped your PR's actual changes with the minor surface differences that I would apply after accepting your PR; however, I would prefer that you retain credit for the lines you changed in `git blame` and that my syntactic differences don't paper them over.

I've attached a file [`41.diff.txt`](https://github.com/myrrlyn/bitvec/files/4104776/41.diff.txt) with the end result of applying your branch and reshaping it. (GitHub requires the `.txt` extension, unfortunately). If you are comfortable with this diff, please overwrite your current PR branch with it. You may also feel free to add your name to [`AUTHORS.txt`](https://github.com/myrrlyn/bitvec/blob/master/AUTHORS.txt) if you wish. Otherwise, I'll merge it as-is and resolve the formatting differences later.

Thank you again for writing this; the new `fields::resize` implementation is much better than my original, and boilerplate reduction in the `store` traits is nice from a maintenance perspective.

**Comment by ImmemorConsultrixContrarie** [comment]
> prefer fixed-size type names in the debug information

Welp, I'd describe writing "usize" as "u64" or "u32" as a mistake, even though it is a renamed unsigned int. Also, you can always `size_of::<usize>()` to get the real size in bytes for format.

And I don't wanna do all this "apply other's changes from diff file" work instead of real coding, so feel free to write whatever changes you did. I'd only suggest adding "unsafe" to the "Sealed" trait and writing "unsafe" in macro implementing it.

**Comment by myrrlyn** [comment]
You're right about `BitMut`; after I realized I couldn't make an actual fake-reference type that commits on drop (Rust forbids `impl Drop for &mut _`, even though it's an affine `!Copy` type), I stopped thinking about how to optimize it and went back to working on the rest of the stdlib-compatibility tasks.

The structure

```rust
struct BitMut<'a, O, T>
where O: BitOrder, T: BitStore {
  _parent: PhantomData<&'a mut BitSlice<O, T>>,
  data: NonNull<T::Access>,
  head: BitIdx<T>,
  bit: bool,
}
```

contains all the information needed to operate, and fits in two words. Thanks for the input!

As the pointer itself is never modified, its slot does not need to be atomic.

**Comment by ImmemorConsultrixContrarie** [comment]
> As the pointer itself is never modified, its slot does not need to be atomic.

IIRC, AtomicPtr<T> has atomic *access* to the data.

**Comment by myrrlyn** [comment]
It does not; it is equivalent to an `AtomicUsize` except that its value is a memory address. `AtomicPtr::load` returns a `*mut T`, which has ordinary non-atomic derefence behavior.

Which is really irritating, because it *does look* like the kind of name you'd give to a pointer with permanent `Relaxed` dereference semantics.

**Comment by myrrlyn** [comment]
I'm going to close this, as I just applied the rollup commit to master.
